Fix beatmap carousel not maintaining selection if currently selected beatmap is updated

peppy opened the pull request:
RFC. I've tested this to work, but adding tests is a bit involved. Want to make sure the solution seems okay before adding them.

- Closes #19328
- [x] Add tests

peppy committed:
Fix beatmap carousel not maintaining selection if currently selected beatmap is updated

LittleEndu commented:
Looking at the code, doesn't this fail when the update to the beatmap is a deletion of the selected difficulty?

peppy commented:
Can you explain what you mean and which line is at fault? I can't read minds.

LittleEndu commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -265,6 +265,38 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
+                            // Best effort matching. We can't use ID because in the update flow a new version will get its own GUID.
+                            bool selectionMatches =
+                                ((IBeatmapMetadataInfo)beatmapInfo.Metadata).Equals(SelectedBeatmapInfo.Metadata)
+                                && beatmapInfo.DifficultyName == SelectedBeatmapInfo.DifficultyName;
```
I meant this. Isn't this true only if the difficulty also exists in the updated beatmap.

I would think if the beatmap set's metadata matches, it's OK to select any difficulty, not necessarily the one that was previously selected.

peppy replied:
Right. A `BeatmapSet`'s metadata is currently just a reference to the first beatmap in the set, so it gets a bit difficult to guarantee anything there.

There are other scenarios such as if the difficulty name is changed (or the title of the beatmap is changed) which will lose selection in a similar way, so if we want to be more inclusive then a fallback selection can be added. See https://github.com/ppy/osu/pull/19409/commits/24d75612e222fe3e055d1f5fabf623727dc81727.

peppy committed:
Always attempt to follow selection, even if difficulty name / metadata change

peppy committed:
Guard against `NewModifiedIndices` being empty

LittleEndu commented:
I want to try this. Should I wait for tests to be added or can I test in prod?

frenzibyte commented:
You should be able to test this in-game by modifying the beatmap however you want then update and see if it selects properly.

peppy commented:
To clarify, you can use the staging/dev environment and download a beatmap, then make a change via realm studio or the editor.

peppy commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -265,6 +265,41 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
             foreach (int i in changes.InsertedIndices)
                 UpdateBeatmapSet(sender[i].Detach());
+
+            if (changes.DeletedIndices.Length > 0)
```
That's not good for the methodology used in this PR. I've asked about this over at realm so we can better understand the scenario (https://github.com/realm/realm-dotnet/discussions/2634#discussioncomment-3277702).

peppy replied:
@frenzibyte I've tried and can't reproduce what you're seeing. Can you share your process for getting into that state? Are you making an edit in the editor or something else?

peppy replied:
Managed to reproduce this, and fixed the case.

peppy committed:
Merge branch 'master' into carousel-maintain-selection-over-update

peppy committed:
Merge branch 'master' into carousel-maintain-selection-over-update

peppy committed:
Change conditional to check for insertions in addition to modifications

It is possible that the import process itself marks the previous
beatmaps as deleted due to an overlap in metadata or otherwise.

peppy commented:
This was incorrectly blocked, should actually be okay to review.

frenzibyte committed:
Merge branch 'master' into carousel-maintain-selection-over-update

frenzibyte committed:
Fix selection fallback path not updated to check inserted indices

frenzibyte committed:
Update comment to include inserted indices

frenzibyte commented:
@peppy Do you still plan to add tests to this PR? It's blocking merge currently.

peppy commented:
will see what i can do

peppy committed:
Add test coverage of selection being retained

peppy commented:
I've added test coverage for this (confirmed to fail on `master`).

frenzibyte committed:
Merge branch 'master' into carousel-maintain-selection-over-update
